keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -271,17 +271,13 @@ example further down).
 
 ### Default publishing
 
-CRDs do not publish default via the OpenAPI spec on the /openapi/v2
-endpoint today, but we have to decide what to do with the new default
-values of built-in types. The kube-openapi generated schemas will
-include them.
-
-We proppose to:
-1. leave CRDs not publishing defaults
-2. filter out defaults in the built-in type schemas too in a first step
-3. but possibly reconsider this behaviour in the future when the
-necessity of defaults for client-side merging (e.g. through kustomize)
-and influences on the ecosystem of default mis-use is better understood.
+Before 1.22, CRDs and build-in types were not exposing their default values through the `/openapi/v2` endpoint but will expose them in Kubernetes v1.22.
+
+These default values can be valuable for generating documentation, and can be required for clients that are trying to merge associative lists with defaulted keys.
```

Comment:
How is merging lists a good use-case, but the one in the paragraph below is not? Do we want clients to do this kind of work with deep knowledge of server behaviour?

For assocciative lists, it would be enough to publish defaults for `list-type=map` items and only for those fields in `x-kubernetes-list-map-keys`.

Reply:
> How is merging lists a good use-case, but the one in the paragraph below is not? Do we want clients to do this kind of work with deep knowledge of server behaviour?

Because, especially for apply, it shows a different intent. Clients sometimes have to merge things for pre-processing (i.e. kustomize) and they can't do that without knowing the default for the key. Automatically specifying the value on the other hand, changes the intent of the person who submits.

> For assocciative lists, it would be enough to publish defaults for list-type=map items and only for those fields in x-kubernetes-list-map-keys.

I don't understand why we wouldn't provide the information to clients/documentation? Knowing the default value is valuable.

Reply:
I miss a discussion here of the pros and cons. That's why I ask these questions, not so much to discuss them in this thread, but to have a clear answer in the enhancement text.

For instance some thoughts:
- publishing defaults makes CRDs and native types different
- can you assume that there is no default when a default is not published?
- admission can do defaulting and this will never be published

Hence, published defaults is best-effort and by design incomplete.

Hence, using defaults for anything other than pure documentation, is questionable. You cannot rely to have them in the schema.

if you are really only interested in documentation, add an automatic sentence to the description in a generation. But those associated lists show that documentation is not the only goal here.

How do you enforce that defaults for your associative lists use-case are part of the schema of native types?

Reply:
> I miss a discussion here of the pros and cons. That's why I ask these questions, not so much to discuss them in this thread, but to have a clear answer in the enhancement text.

Sounds good!

> publishing defaults makes CRDs and native types different

I don't get that? We would publish defaults for both?

> Hence, using defaults for anything other than pure documentation, is questionable. You cannot rely to have them in the schema.

Depends what we call documentation, but I'm also including anything that helps author resources, including editor auto-completion, web UIs, etc.

> if you are really only interested in documentation, add an automatic sentence to the description through schema generation. But those associated lists show that documentation is not the only goal here.

I think that having defaults as a programmatically understandable artifacts helps the UI.

Reply:
> I don't get that? We would publish defaults for both?

We have defaults in native types that cannot be expressed with OpenAPI defaulting.

Reply:
I don't know if it's a strong enough reason not to expose the static defaults none-the-less. They do provide additional information, I don't see a good reason to hide that.

Reply:
it depeneds on the guarantee: if you say they are best effort and some default might be missing, it's probably fine. If somebody builds tooling that breaks when a default is missing, that's not a good outcome.

Reply:
OK Yeah, I agree with that. They have to be guaranteed only for non-required associative keys. Other than that, it's definitely best effort. I don't know if the OpenAPI spec specifically makes stronger guarantees than this though, so I don't think that's a mis-use of the `default` field.